### PR TITLE
provider/kubernetes: Register docker registry credentials.

### PR DESCRIPTION
**Pull request by lwander**
This PR achieves 2 things.

1. The Kubernetes user can define a list of docker registry accounts whose credentials and endpoint will be registered with as a secret with the Kubernetes master. You can read more about how this works [here](http://kubernetes.io/v1.0/docs/user-guide/images.html). The user can also provide an optional list of namespaces (if omitted, every namespace is used) that this registry can be used to deploy images to. The motivation here is to allow users to configure repositories with images specifically for test, prod, staging, etc, if they want to.

2. While writing the mock tests, I was frustrated with the fact that `KubernetesUtil` was non-static, and had a subset of kubernetes API functionality built in. I took this functionality out, made the remainder static, and placed it into `KubernetesApiAdaptor`. I will now do all API calls through this class. If we need to plug in a separate Kubernetes API it will be easy to abstract and then subclass `KubernetesApiAdaptor`.

### CODE REVIEW

**Comment by lwander** [comment]
@duftler

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy [hunk 1]
```diff
@@ -31,9 +31,27 @@ public class DockerRegistryNamedAccountCredentials implements AccountCredentials
   public DockerRegistryNamedAccountCredentials(String accountName, String environment, String accountType,
                                                String address, String username, String password, String email,
                                                List<String> repositories, List<String> requiredGroupMembership) {
+    if (!accountName || accountName.length() == 0) {
```
Comment (duftler) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy]: Can just be `if (!accountName)`.

http://groovy-lang.org/semantics.html#_strings

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy [hunk 2]
```diff
@@ -31,9 +31,27 @@ public class DockerRegistryNamedAccountCredentials implements AccountCredentials
   public DockerRegistryNamedAccountCredentials(String accountName, String environment, String accountType,
```
Comment (duftler) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy]: Can drop semi-colons in this file.

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy [hunk 3]
```diff
@@ -31,9 +31,27 @@ public class DockerRegistryNamedAccountCredentials implements AccountCredentials
...
                                                List<String> repositories, List<String> requiredGroupMembership) {
+    if (!accountName || accountName.length() == 0) {
+      throw new IllegalArgumentException("Docker Registry account must be provided with a name.");
+    }
     this.accountName = accountName;
     this.environment = environment;
     this.accountType = accountType;
+
+    if (!address || address.length() == 0) {
+      throw new IllegalArgumentException("Docker Registry account " + accountName + " must provide an endpoint address.");
```
Comment (duftler) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy]: Can use Groovy string magic here.

File: clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy [hunk 4]
```diff
@@ -48,6 +66,24 @@ public class DockerRegistryNamedAccountCredentials implements AccountCredentials
...
+        "" :
+      "";
+  }
+
+  public String getEmail() {
+    return email;
+  }
+
+  public String getV2Endpoint() {
+    return address + "/v2";
```
Comment (duftler) [inline clouddriver-docker/src/main/groovy/com/netflix/spinnaker/clouddriver/docker/registry/security/DockerRegistryNamedAccountCredentials.groovy]: `return "$address/v2"`

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidator.groovy [hunk 5]
```diff
@@ -19,6 +19,8 @@ package com.netflix.spinnaker.clouddriver.kubernetes.deploy.validators
 import com.netflix.spinnaker.clouddriver.deploy.DescriptionValidator
 import com.netflix.spinnaker.clouddriver.kubernetes.deploy.description.CloneKubernetesAtomicOperationDescription
 import com.netflix.spinnaker.clouddriver.kubernetes.KubernetesOperation
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/CloneKubernetesAtomicOperationValidator.groovy]: Sort-order is off here.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/health/KubernetesHealthIndicator.groovy [hunk 6]
```diff
@@ -69,11 +69,11 @@ class KubernetesHealthIndicator implements HealthIndicator {
         // This verifies that the specified credentials are sufficient to
         // access the referenced Kubernetes master endpoint.
         kubernetesCredentials.getNamespaces().each { namespace ->
-          Namespace res = kubernetesCredentials.client.namespaces().withName(namespace).get();
+          Namespace res = kubernetesCredentials.apiAdaptor.getNamespace(namespace)
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/health/KubernetesHealthIndicator.groovy]: This does look much nicer.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/provider/agent/KubernetesServerGroupCachingAgent.groovy [hunk 7]
```diff
@@ -32,6 +32,7 @@ import com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentia
 import groovy.util.logging.Slf4j
 import io.fabric8.kubernetes.api.model.Pod
 import io.fabric8.kubernetes.api.model.ReplicationController
+import org.springframework.beans.factory.annotation.Autowired
```
Comment (duftler) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/provider/agent/KubernetesServerGroupCachingAgent.groovy]: Extraneous?

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperationSpec.groovy [hunk 8]
```diff
@@ -18,15 +18,21 @@ package com.netflix.spinnaker.clouddriver.kubernetes.deploy.ops
...
 import com.netflix.spinnaker.clouddriver.kubernetes.deploy.KubernetesUtil
 import com.netflix.spinnaker.clouddriver.kubernetes.deploy.description.KubernetesContainerDescription
 import com.netflix.spinnaker.clouddriver.kubernetes.deploy.description.KubernetesResourceDescription
 import com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials
+import com.netflix.spinnaker.clouddriver.security.AccountCredentialsRepository
 import io.fabric8.kubernetes.api.model.ObjectMeta;
 import io.fabric8.kubernetes.api.model.IntOrString
 import io.fabric8.kubernetes.api.model.ReplicationController
 import io.fabric8.kubernetes.api.model.ReplicationControllerList
+import io.fabric8.kubernetes.api.model.Secret
```
Comment (duftler) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperationSpec.groovy]: Extraneous?

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperationSpec.groovy [hunk 9]
```diff
@@ -18,15 +18,21 @@ package com.netflix.spinnaker.clouddriver.kubernetes.deploy.ops
...
 import io.fabric8.kubernetes.api.model.Service
 import io.fabric8.kubernetes.api.model.ServiceList
 import io.fabric8.kubernetes.api.model.ServicePort
```
Comment (duftler) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperationSpec.groovy]: I think `KubernetesClient` is extraneous as well.

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperationSpec.groovy [hunk 10]
```diff
@@ -18,15 +18,21 @@ package com.netflix.spinnaker.clouddriver.kubernetes.deploy.ops
 
 import com.netflix.spinnaker.clouddriver.data.task.Task
 import com.netflix.spinnaker.clouddriver.data.task.TaskRepository
+import com.netflix.spinnaker.clouddriver.docker.registry.security.DockerRegistryCredentials
```
Comment (duftler) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperationSpec.groovy]: Extraneous?

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperationSpec.groovy [hunk 11]
```diff
@@ -91,8 +97,17 @@ class DeployKubernetesAtomicOperationSpec extends Specification {
     servicePortMock = Mock(ServicePort)
     metadataMock = Mock(ObjectMeta)
     intOrStringMock = Mock(IntOrString)
+    accountCredentialsRepositoryMock = Mock(AccountCredentialsRepository)
+
+    DOCKER_REGISTRY_ACCOUNTS.forEach({ account ->
```
Comment (duftler) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/DeployKubernetesAtomicOperationSpec.groovy]: Are the enclosing parens needed here?

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidatorSpec.groovy [hunk 12]
```diff
@@ -37,7 +49,21 @@ class StandardKubernetesAttributeValidatorSpec extends Specification {
     accountCredentialsProvider = new DefaultAccountCredentialsProvider(credentialsRepo)
     def namedAccountCredentialsMock = Mock(KubernetesNamedAccountCredentials)
     namedAccountCredentialsMock.getName() >> ACCOUNT_NAME
-    namedAccountCredentialsMock.getCredentials() >> new KubernetesCredentials(null, null)
+    def apiMock = Mock(KubernetesApiAdaptor)
```
Comment (duftler) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/validators/StandardKubernetesAttributeValidatorSpec.groovy]: Could extract this routine into a helper.

**Comment by duftler** [comment]
LGTM. Feel free to self-merge after addressing comments and rebasing.

**Pull request merged**